src/wallet/wallet.cpp

```diff
@@ -417,6 +417,12 @@ bool CWallet::ChangeWalletPassphrase(const SecureString& strOldWalletPassphrase,
 
 void CWallet::SetBestChain(const CBlockLocator& loc)
 {
+    LOCK(cs_wallet); //nWalletMaxVersion
+    unsigned int keypool_min = GetArg("-keypoolmin", DEFAULT_KEYPOOL_MIN);
+    if (IsHDEnabled() && ((CanSupportFeature(FEATURE_HD_SPLIT) && setInternalKeyPool.size() < keypool_min) || (setExternalKeyPool.size() < keypool_min))) {
```

Comment:
Thread https://github.com/bitcoin/bitcoin/pull/10882#discussion_r131245392

Posted a question in that commit, but I think this seems basically right. Few things

- I don't think this change necessarily needs to be part of this PR if it will delay the merge. Could be done as a followup.
- It seems like it would be a little safer (though maybe equivalent) and more logical to change `m_update_best_block` from true to false in `AddToWalletIfInvolvingMe` between `TopUpKeyPool` and `ShutdownIfKeypoolCritical`, since similar checks are being done there, and it would make it obvious that if there's any dip in keypool size while transactions are being processed, the flag will flip.
- When I talked about this with Alex he wanted to keep `-bypasskeypoolcritical` even if `-keypool_critical=0` can be made safe & equivalent, because it makes purpose of the flag more explicit.
- Would be good to use the HasUnusedKeys helper to simplify the logic.